[joellow88] iP

joellow88 opened the pull request:
# ip - DUKE
## Your personal assistant

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://github.com/joellow88/ip/releases/tag/A-Jar)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :grinning:

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

#

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
    Duke d = new Duke();
    d.ui.showStartingDialogue();
    
    while (d.isRunning) {
        d.isRunning = d.ui.run(d.tasks);
        d.storage.saveData(d.tasks);
    }
```

Commit:
Add Gradle support

joellow88 committed:
Completed Level-1 with Greet, Echo and Exit functionality

joellow88 committed:
Completed Level-2
Included functionality to add objects to list and print list

joellow88 committed:
Added Task.java
Refactoring of Duke.java to use Task.java is not complete

joellow88 committed:
Updated Task.java with printTask() method
Added mark and unmark functionality to Duke.java

joellow88 committed:
Completed Level-4
Refactored Task to be an abstract class
Added subclasses Todo, Deadline, Event
Added new response in Duke to catch non-commands

joellow88 committed:
Updated test cases up to Level-4.
Error with test cases failing even though EXPECTED.TXT and ACTUAL.TXT are identical.
To remedy.

joellow88 committed:
Updated input.txt and EXPECTED.TXT with test cases up to Level 4
Corrected runtest.bat so testing script runs correctly

joellow88 committed:
Added basic error handling.
Met minimal requirements for Level-5 to handle undefined commands and empty todo.

joellow88 committed:
Implemented delete for Level-6

joellow88 committed:
Refactored error handling using Regex

joellow88 committed:
Added new test cases to include delete, and error handling

joellow88 committed:
Refactor UI and command handling to use UiHandler and CommandHandler
Update EXPECTED.TXT to match input

joellow88 committed:
Refactor Duke, UiHandler and CommandHandler to use instance methods instead of class methods

joellow88 committed:
Add Parser to parse user input.
Add Command to send command from parser to CommandHandler

joellow88 committed:
Add TaskList.java
Refactored Duke, UiHandler and CommandHandler to use TaskList

joellow88 committed:
Refactor all files into package duke
Update test script to compile package duke

joellow88 committed:
Edit runtest.bat

joellow88 committed:
Edit runtest.bat so it works properly with package duke

joellow88 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

joellow88 committed:
Edit build.gradle
Set up IDE to build and run using gradle

joellow88 committed:
Add test cases for Command and CommandHandler
Refactor TaskList to return reference to itself when adding and deleting tasks

joellow88 committed:
Add equals() and hashcode() methods for Command
Add JUnit tests for Parser

joellow88 committed:
Add Level-7
Able to store TaskList in a .txt file to be loaded for later use
Able unable to store done and undone task status

joellow88 committed:
Add /data/ to .gitignore

joellow88 committed:
Add saving and restoring of task status

joellow88 committed:
Complete Level-8
Refactor Deadline, Event to use LocalDateTime
Add DateTimeParser
Update JUnit test cases

joellow88 committed:
Merge branch 'branch-Level-7'

joellow88 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/duke/CommandHandler.java
#	src/main/java/duke/Deadline.java
#	src/main/java/duke/Parser.java

joellow88 committed:
Add MANIFEST.MF

joellow88 committed:
Add JavaDocs.

joellow88 committed:
Amend code to comply with coding standards.

Coding standards found at
https://se-education.org/guides/conventions/java/intermediate.html

joellow88 committed:
Add find task

Duke does not support searching for tasks in the task list.

It is useful to have a function to search for tasks in a large task list.

Adding a find command, corresponding to a search function, achieves this functionality.

joellow88 committed:
Merge branch 'branch-A-JavaDoc'

joellow88 committed:
Merge branch 'branch-A-CodingStandard'

joellow88 committed:
Merge branch 'branch-Level-9'

joellow88 committed:
Refactor code to fit the coding standard

francisyzy commented on src/main/java/duke/CommandHandler.java:
```diff
@@ -0,0 +1,167 @@
+            //Fallthrough
+        case "find":
+            return findTasks(command.getArguments().get(0), tasks);
+        }
```
Hey, I think you're missing the default case!

francisyzy commented on src/main/java/duke/DateTimeParser.java:
```diff
@@ -0,0 +1,56 @@
+            LocalDateTime dt = date.atTime(LocalTime.of(0, 0));
+            return dt;
+        } catch (DateTimeParseException e) { }
+        throw new DateTimeParseException("Unable to parse datetime", dateTime, 0);
```
Can you group all of the try catch into one block? Not sure what would be the best practice here.

francisyzy commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,60 @@
+package duke;
+
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
```
Is this an unused import?

francisyzy commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,68 @@
+     * @param isDone The done status of the Event task.
+     * @param startTime The start time of the Event task as a String.
+     * @param endTime The end time of the Event task as a String.
+     */
```
Great work on the documentation!

francisyzy commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,63 @@
+     * Constructs a Command with the specified description and any necessary arguments.
+     * 
+     * @param description The Command description (eg: todo, list, mark, etc)
+     * @param arguments The necessary arguments for the command (eg: command "deadline" can have argument "11/02/2023 1200"
```
Would like to check if it is possible for comments to be wrapped. This line is getting long

francisyzy commented on src/main/java/duke/CommandHandler.java:
```diff
@@ -0,0 +1,167 @@
+     * @param tasks A TaskList containing the current list of tasks in Duke.
+     *              If required, the TaskList will be modified in place as a side effect of the method call.
+     * @return The response that Duke will display upon execution of the command.
+     */
```
I like the way you do your JavaDocs!

francisyzy reviewed (commented):
Great job documenting your code! There are a few minor issues with the styling but otherwise it LGTM. Do try and use [checkstyle](https://nus-cs2103-ay2223s2.github.io/website/admin/ip-w4.html#a-checkstyle) as it will help catch minor styling issues.

joellow88 committed:
Add GUI to Duke
Rename UiHandler to CliHandler

joellow88 committed:
Refactor Command to use Varargs

joellow88 committed:
Merge branch 'branch-Level-10'

joellow88 committed:
Merge branch 'branch-A-Varargs'
Resolve merge conflicts
# Conflicts:
#	src/main/java/duke/CommandHandler.java

joellow88 committed:
Better look for Duke

The interface for Duke leaves no space between DialogBox.
There is also an awkward misalignment between the Images and Labels.

This gives the Duke GUI a cluttered look and feel.

Adding padding to the DialogBox will allow space between DialogBoxes.
Adding padding to Labels will allow space between the Label and Image
while correcting the awkward misalignment.

Let's insert padding of (top: 20, right: 0, bottom: 20, left: 0) to DialogBoxes.
Let's insert padding of (0, 10, 0, 10) to Labels.

Using Padding is preferable to using borders as we can still add custom
borders in future improvements to the GUI. This may be used to implement
UI improvements like speech bubbles, or line separators.

joellow88 committed:
New look for Duke and User

The images for Duke and User are the default images in the tutorial.

The default images are not very appealing and not suitable for this app.
The images also remain the same regardless of the state of the app.

Lets change the images to be more appealing.
Lets also dynamically alter the user image shown depending on
how many tasks are in the current task list.

This will provide a more enjoyable experience for the end user.

joellow88 committed:
Complete C-NaturalDates

Duke can only understand dates in a few very specific formats.

This is unintuitive for the user and many more natural dates cannot be entered.

Lets update DateTimeParser to allow Duke to understand many more datetime formats.
We add a list of acceptable dates and time formats into separate arrays.
Duke then tries to format a given input against any of those formats.
If none match, Duke combines the dates and time formats and tries to format the inputs again.

This allows us to easily add new supported date time formats in the future.

joellow88 committed:
Doing A-Assertions.

Add assertions to Duke.

joellow88 committed:
Complete A-Assertions
Added assertions to CommandHandler, Parser and Storage.
Enabled assertions in build.gradle

joellow88 committed:
Implement exceptions for reaching default branch in CommandHandler and Parser.

joellow88 committed:
Include comments to explain the empty catch block.

joellow88 committed:
Merge pull request #1 from joellow88/branch-C-NaturalDates

Complete C-NaturalDates

joellow88 committed:
Merge branch 'master' of https://github.com/joellow88/ip

* 'master' of https://github.com/joellow88/ip:
  Complete C-NaturalDates

joellow88 committed:
Merge branch 'master' into branch-A-Assertions

joellow88 committed:
Merge branch 'master' into branch-A-CodeQuality
Resolve merge conflicts

# Conflicts:
#	src/main/java/duke/DateTimeParser.java

joellow88 committed:
Merge pull request #2 from joellow88/branch-A-Assertions

Branch a assertions

joellow88 committed:
Merge branch 'master' of https://github.com/joellow88/ip

* 'master' of https://github.com/joellow88/ip:
  Complete A-Assertions Added assertions to CommandHandler, Parser and Storage. Enabled assertions in build.gradle
  Doing A-Assertions.

joellow88 committed:
Merge branch 'master' into branch-A-CodeQuality
Resolve merge conflicts

# Conflicts:
#	src/main/java/duke/CommandHandler.java
#	src/main/java/duke/Parser.java

joellow88 committed:
Merge pull request #3 from joellow88/branch-A-CodeQuality

Branch a code quality

joellow88 committed:
Merge branch 'master' of https://github.com/joellow88/ip

* 'master' of https://github.com/joellow88/ip:
  Include comments to explain the empty catch block.
  Implement exceptions for reaching default branch in CommandHandler and Parser.

joellow88 committed:
Fixed bugs with assertions in Parser

joellow88 committed:
Rename DateTimeParser to DateTimeHandler

Refactor DateTimeHandler to use instance methods
instead of static methods to parse an input string.

Add new case of error handling for when
user inputs a date in the correct format but does not exist.

joellow88 committed:
Fix bugs in parsing some DateTime edge cases (29 Feb, 31 Mar)

joellow88 committed:
Add UX improvement

Press UP arrow key enters command history in user input field

joellow88 committed:
Improve find to show the index of the task in the TaskList

joellow88 committed:
Add User Guide

joellow88 committed:
Add representative screenshot to user guide

joellow88 committed:
Update README.md
